File: lib/Mach1EncodeEmscripten.js
Module['onInited'] = function() {
};

Module['onRuntimeInitialized'] = function() {
    console.log('loaded Mach1Encode');

	Module['Mach1Encode'] = function() {
		this.Mach1EncodeInputModeType = {
			"Mach1EncodeInputModeMono": (Module).Mach1EncodeInputModeType.Mach1EncodeInputModeMono,
			"Mach1EncodeInputModeStereo": (Module).Mach1EncodeInputModeType.Mach1EncodeInputModeStereo,
			"Mach1EncodeInputModeQuad": (Module).Mach1EncodeInputModeType.Mach1EncodeInputModeQuad,
			"Mach1EncodeInputModeAFormat": (Module).Mach1EncodeInputModeType.Mach1EncodeInputModeAFormat,
			"Mach1EncodeInputModeBFormat": (Module).Mach1EncodeInputModeType.Mach1EncodeInputModeBFormat,
		};

		this.Mach1EncodeOutputModeType = {
			"Mach1EncodeOutputMode4Ch": (Module).Mach1EncodeOutputModeType.Mach1EncodeOutputMode4Ch,
			"Mach1EncodeOutputMode8Ch": (Module).Mach1EncodeOutputModeType.Mach1EncodeOutputMode8Ch,
		};

		this.Mach1DecodeAlgoType = {
			"Mach1DecodeAlgoSpatial": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoSpatial,
			"Mach1DecodeAlgoAltSpatial": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoAltSpatial,
			"Mach1DecodeAlgoHorizon": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoHorizon,
			"Mach1DecodeAlgoHorizonPairs": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoHorizonPairs,
			"Mach1DecodeAlgoSpatialPairs": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoSpatialPairs,
		};

		this.delete = function() {
			if (this._m1obj) {
				this._m1obj.delete();
				this._m1obj = null;
			}
		};

		this.getPoints = function() {
			if (this._m1obj) {
				var arr = this._m1obj.getPoints();
				var points = [];
				for (var i = 0; i < arr.size(); i++) {
					points[i] = {
						x: arr.get(i).x,
						y: arr.get(i).y,
						z: arr.get(i).z
					};
				}
				arr.delete();
				arr = null;
				
				return points
			}
		};

		this.getGains = function() {
			if (this._m1obj) {
				var arr = this._m1obj.getGains();
				var gains = [];
				for (var i = 0; i < arr.size(); i++) {
					gains[i] = [];
					var _arr = arr.get(i);
					for (var j = 0; j < _arr.size(); j++) {
						gains[i][j] = _arr.get(j);
					}
					_arr.delete();
					_arr = null;
				}
				arr.delete();
				arr = null;

				return gains;
			}
		};

		this.getResultingVolumesDecoded = function(decodeType, decodeResult) {
			if (this._m1obj) {
				var vec = new (Module).VectorFloat();
				for (var i = 0; i < decodeResult.length; i++) {
					vec.push_back(decodeResult[i]);
				}
				
				var arr = this._m1obj.getResultingVolumesDecoded(decodeType, vec);
				var volumes = [];
				for (var i = 0; i < arr.size(); i++) {
					volumes[i] = arr.get(i);
				}
				arr.delete();
				arr = null;

				vec.delete();
				vec = null;

				return volumes;
			}
		};

		this.getPointsNames = function() {
			if (this._m1obj) {
				var arr = this._m1obj.getPointsNames();
				var pointsNames = [];
				for (var i = 0; i < arr.size(); i++) {
					pointsNames[i] = arr.get(i);
				}
				arr.delete();
				arr = null;

				return pointsNames;
			}
		};

		this.getGainsForInputChannelNamed = function(channelName) {
			if (this._m1obj) {
				return this._m1obj.getGainsForInputChannelNamed(channelName);
			}
		};

		this.generatePointResults = function() {
			if (this._m1obj) {
				this._m1obj.generatePointResults();
			}
		};

		this.getPointsCount = function() {
			if (this._m1obj) {
				return this._m1obj.getPointsCount();
			}
		};

		this.setInputMode = function(inputMode) {
			if (this._m1obj) {
				this._m1obj.setInputMode(inputMode);
			}
		};

		this.setInputMode = function(inputMode) {
			if (this._m1obj) {
				this._m1obj.setInputMode(inputMode);
			}
		};

		this.setOutputMode = function(outputMode) {
			if (this._m1obj) {
				this._m1obj.setOutputMode(outputMode);
			}
		};

		this.setRotation = function(rotation) {
			if (this._m1obj) {
				this._m1obj.setAzimuthDegrees(rotation);
			}
		};

    this.setAzimuth = function(azimuth) {
      if (this._m1obj) {
        this._m1obj.setAzimuth(azimuth);
      }
    };

    this.setAzimuthDegrees = function(azimuth) {
      if (this._m1obj) {
        this._m1obj.setAzimuthDegrees(azimuth);
      }
    };

     this.setAzimuthRadians = function(azimuth) {
      if (this._m1obj) {
        this._m1obj.setAzimuthRadians(azimuth);
      }
    };

		this.setDiverge = function(diverge) {
			if (this._m1obj) {
				this._m1obj.setDiverge(diverge);
			}
		};

		this.setPitch = function(pitch) {
			if (this._m1obj) {
				this._m1obj.setElevationDegrees(pitch);
			}
		};

    this.setElevation = function(elevation) {
      if (this._m1obj) {
        this._m1obj.setElevation(elevation);
      }
    };

    this.setElevationDegrees = function(elevation) {
      if (this._m1obj) {
        this._m1obj.setElevationDegrees(elevation);
      }
    };

    this.setElevationRadians = function(elevation) {
      if (this._m1obj) {
        this._m1obj.setElevationRadians(elevation);
      }
    };

    this.setIsotropicEncode = function(isotropicEncode) {
      if (this._m1obj) {
        this._m1obj.setIsotropicEncode(isotropicEncode);
      }
    };

		this.setStereoRotate = function(stereoRotate) {
			if (this._m1obj) {
				this._m1obj.setOrbitRotationDegrees(stereoRotate);
			}
		};

    this.setOrbitRotation = function(orbitRotation) {
      if (this._m1obj) {
        this._m1obj.setOrbitRotation(orbitRotation);
      }
    };

    this.setOrbitRotationDegrees = function(orbitRotation) {
      if (this._m1obj) {
        this._m1obj.setOrbitRotationDegrees(orbitRotation);
      }
    };

    this.setOrbitRotationRadians = function(orbitRotation) {
      if (this._m1obj) {
        this._m1obj.setOrbitRotationRadians(orbitRotation);
      }
    };

		this.setStereoSpread = function(stereoSpread) {
			if (this._m1obj) {
				this._m1obj.setStereoSpread(stereoSpread);
			}
		};

		this.setAutoOrbit = function(autoOrbit) {
			if (this._m1obj) {
				this._m1obj.setAutoOrbit(autoOrbit);
			}
		};

		this._m1obj = new(Module).Mach1EncodeInternal();
	};

	Module['onInited']();
};

File: lib/Mach1EncodeEmscripten.cpp
#include <emscripten/bind.h>

#include "Mach1EncodeCAPI.h"
#include "Mach1DecodeCAPI.h"
#include "external/cpp/Mach1Encode.cpp"

using namespace emscripten;

EMSCRIPTEN_BINDINGS(Mach1Encode) {
    enum_<Mach1EncodeInputModeType>("Mach1EncodeInputModeType")
        .value("Mach1EncodeInputModeMono", Mach1EncodeInputModeMono)
        .value("Mach1EncodeInputModeStereo", Mach1EncodeInputModeStereo)
        .value("Mach1EncodeInputModeQuad", Mach1EncodeInputModeQuad)
        .value("Mach1EncodeInputModeAFormat", Mach1EncodeInputModeAFormat)
        .value("Mach1EncodeInputModeBFormat", Mach1EncodeInputModeBFormat)
        ;
		
    enum_<Mach1EncodeOutputModeType>("Mach1EncodeOutputModeType")
        .value("Mach1EncodeOutputMode4Ch", Mach1EncodeOutputMode4Ch)
        .value("Mach1EncodeOutputMode8Ch", Mach1EncodeOutputMode8Ch)
        ;

	enum_<Mach1DecodeAlgoType>("Mach1DecodeAlgoType")
        .value("Mach1DecodeAlgoSpatial", Mach1DecodeAlgoSpatial)
        .value("Mach1DecodeAlgoAltSpatial", Mach1DecodeAlgoAltSpatial)
        .value("Mach1DecodeAlgoHorizon", Mach1DecodeAlgoHorizon)
        .value("Mach1DecodeAlgoHorizonPairs", Mach1DecodeAlgoHorizonPairs)
        .value("Mach1DecodeAlgoSpatialPairs", Mach1DecodeAlgoSpatialPairs)
        ;

	register_vector<float>("VectorFloat");
	register_vector<std::string>("VectorString");
	register_vector<Mach1Point3D>("VectorMach1Point3D");
	register_vector<std::vector<float>>("VectorVectorFloat");

    value_object<Mach1Point3D>("Mach1Point3D")
		.field("x", &Mach1Point3D::x)
		.field("y", &Mach1Point3D::y)
		.field("z", &Mach1Point3D::z)
        ;

    class_<Mach1Encode>("Mach1EncodeInternal")
        .constructor<>()
        .function("getPoints", &Mach1Encode::getPoints)
        .function("getGains", &Mach1Encode::getGains)
		.function("getPointsNames", &Mach1Encode::getPointsNames)
		.function("getGainsForInputChannelNamed", &Mach1Encode::getGainsForInputChannelNamed)

		.function("generatePointResults", &Mach1Encode::generatePointResults)
		.function("getPointsCount", &Mach1Encode::getPointsCount)

		.function("getResultingVolumesDecoded", &Mach1Encode::getResultingVolumesDecoded)

		.function("setInputMode", &Mach1Encode::setInputMode)
		.function("setOutputMode", &Mach1Encode::setOutputMode)
		.function("setRotation", &Mach1Encode::setRotation)
        .function("setAzimuthDegrees", &Mach1Encode::setAzimuthDegrees)
        .function("setAzimuthRadians", &Mach1Encode::setAzimuthRadians)
        .function("setAzimuth0to1", &Mach1Encode::setAzimuth0to1)

		.function("setDiverge", &Mach1Encode::setDiverge)
        .function("setPitch", &Mach1Encode::setPitch)
        .function("setElevationDegrees", &Mach1Encode::setElevationDegrees)
        .function("setElevationRadians", &Mach1Encode::setElevationRadians)
        .function("setElevation0to1", &Mach1Encode::setElevation0to1)
        .function("setIsotropicEncode", &Mach1Encode::setIsotropicEncode)

        .function("setStereoRotate", &Mach1Encode::setStereoRotate)
        .function("setOrbitRotationDegrees", &Mach1Encode::setOrbitRotationDegrees)
        .function("setOrbitRotationRadians", &Mach1Encode::setOrbitRotationRadians)
        .function("setOrbitRotation0to1", &Mach1Encode::setOrbitRotation0to1)

        .function("setStereoSpread", &Mach1Encode::setStereoSpread)
        .function("setAutoOrbit", &Mach1Encode::setAutoOrbit)
        ;
}
